subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -271,25 +307,33 @@ private void assertValidTaskConfiguration(@Nullable Configuration configuration)
     }
 
 
-    private static AttributeMatchDetails match(Attribute<?> actualAttribute, Object actualValue, AttributeContainer requestedAttributes) {
+    private AttributeMatchDetails match(Attribute<?> actualAttribute, @Nullable Object actualValue, AttributeContainer requestedAttributes) {
         for (Attribute<?> requested : requestedAttributes.keySet()) {
             Object requestedValue = requestedAttributes.getAttribute(requested);
             if (requested.getName().equals(actualAttribute.getName())) {
                 // found an attribute with the same name, but they do not necessarily have the same type
                 if (requested.equals(actualAttribute)) {
-                    if (actualValue.equals(requestedValue)) {
-                        return new AttributeMatchDetails(MatchType.requested, requested, requestedValue);
+                    if (Objects.equals(actualValue, requestedValue)) {
+                        return new AttributeMatchDetails(MatchType.REQUESTED, requested, requestedValue);
                     }
                 } else {
                     // maybe it matched through coercion
-                    if (actualValue.toString().equals(requestedValue.toString())) {
-                        return new AttributeMatchDetails(MatchType.requested, requested, requestedValue);
+                    Object actualString = actualValue != null ? actualValue.toString() : null;
+                    Object requestedString = requestedValue != null ? requestedValue.toString() : null;
+                    if (Objects.equals(actualString, requestedString)) {
+                        return new AttributeMatchDetails(MatchType.REQUESTED, requested, requestedValue);
                     }
                 }
-                return new AttributeMatchDetails(MatchType.different_value, requested, requestedValue);
+                // TODO report "compatible" vs "incompatible" different values (MatchType.INCOMPATIBLE)
+                return new AttributeMatchDetails(MatchType.DIFFERENT_VALUE, requested, requestedValue);
             }
         }
-        return new AttributeMatchDetails(MatchType.not_requested, null, null);
+        return new AttributeMatchDetails(MatchType.NOT_REQUESTED, null, null);
+    }
+
+    private enum SelectionReasonOutput {
+        DEPENDENCY,
+        VARIANT,
     }
 
     private static class AttributeMatchDetails {
```

Comment:
Considering this class is already changing significantly here, I'd extract all these inner types out to top-level classes in the `diagnostics.internal.dependencies` package.